File: zak/core/audit/events.py
"""
ZAK Audit Layer — typed event definitions and structured logging.

All events carry: event_type, agent_id, tenant_id, trace_id, timestamp.
"""

from __future__ import annotations

from datetime import datetime, timezone
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class AuditEventType(str, Enum):
    AGENT_STARTED = "agent.started"
    AGENT_COMPLETED = "agent.completed"
    AGENT_FAILED = "agent.failed"
    TOOL_CALLED = "agent.tool_called"
    TOOL_RESULT = "agent.tool_result"
    DECISION_MADE = "agent.decision"
    POLICY_ALLOWED = "policy.allowed"
    POLICY_BLOCKED = "policy.blocked"
    GRAPH_READ = "sif.graph_read"
    GRAPH_WRITE = "sif.graph_write"
    HUMAN_APPROVAL_REQUESTED = "governance.approval_requested"
    HUMAN_APPROVAL_GRANTED = "governance.approval_granted"
    HUMAN_APPROVAL_DENIED = "governance.approval_denied"


class AuditEvent(BaseModel):
    """Immutable structured audit log entry."""
    event_type: AuditEventType
    agent_id: str
    tenant_id: str
    trace_id: str
    timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    payload: dict[str, Any] = Field(default_factory=dict)


class AgentStartedEvent(AuditEvent):
    event_type: AuditEventType = AuditEventType.AGENT_STARTED


class AgentCompletedEvent(AuditEvent):
    event_type: AuditEventType = AuditEventType.AGENT_COMPLETED
    success: bool = True
    duration_ms: float = 0.0


class AgentFailedEvent(AuditEvent):
    event_type: AuditEventType = AuditEventType.AGENT_FAILED
    error: str = ""


class PolicyBlockedEvent(AuditEvent):
    event_type: AuditEventType = AuditEventType.POLICY_BLOCKED
    action: str = ""
    reason: str = ""


class ToolCalledEvent(AuditEvent):
    event_type: AuditEventType = AuditEventType.TOOL_CALLED
    tool_name: str = ""
    input_summary: str = ""


class GraphWriteEvent(AuditEvent):
    event_type: AuditEventType = AuditEventType.GRAPH_WRITE
    node_type: str = ""
    node_id: str = ""
    operation: str = "upsert"

File: adk/go/pkg/audit/events.go
// Package audit provides typed event definitions and structured logging
// for the ZAK agent audit trail.
//
// All events carry: EventType, AgentID, TenantID, TraceID, Timestamp.
package audit

import "time"

// EventType identifies the category of audit event (13 types).
type EventType string

const (
	AgentStarted           EventType = "agent.started"
	AgentCompleted         EventType = "agent.completed"
	AgentFailed            EventType = "agent.failed"
	ToolCalled             EventType = "agent.tool_called"
	ToolResult             EventType = "agent.tool_result"
	DecisionMade           EventType = "agent.decision"
	PolicyAllowed          EventType = "policy.allowed"
	PolicyBlocked          EventType = "policy.blocked"
	GraphRead              EventType = "sif.graph_read"
	GraphWrite             EventType = "sif.graph_write"
	HumanApprovalRequested EventType = "governance.approval_requested"
	HumanApprovalGranted   EventType = "governance.approval_granted"
	HumanApprovalDenied    EventType = "governance.approval_denied"
)

// Event is the base immutable structured audit log entry.
type Event struct {
	EventType EventType              `json:"event_type"`
	AgentID   string                 `json:"agent_id"`
	TenantID  string                 `json:"tenant_id"`
	TraceID   string                 `json:"trace_id"`
	Timestamp time.Time              `json:"timestamp"`
	Payload   map[string]interface{} `json:"payload,omitempty"`
}

// NewEvent creates a new audit event with the current timestamp.
func NewEvent(eventType EventType, agentID, tenantID, traceID string) *Event {
	return &Event{
		EventType: eventType,
		AgentID:   agentID,
		TenantID:  tenantID,
		TraceID:   traceID,
		Timestamp: time.Now().UTC(),
		Payload:   make(map[string]interface{}),
	}
}

// WithPayload adds key-value pairs to the event payload.
func (e *Event) WithPayload(key string, value interface{}) *Event {
	e.Payload[key] = value
	return e
}

// AgentStartedEvent creates an agent.started event.
func AgentStartedEvent(agentID, tenantID, traceID string) *Event {
	return NewEvent(AgentStarted, agentID, tenantID, traceID)
}

// AgentCompletedEvent creates an agent.completed event with success and duration.
func AgentCompletedEvent(agentID, tenantID, traceID string, success bool, durationMs float64) *Event {
	e := NewEvent(AgentCompleted, agentID, tenantID, traceID)
	e.Payload["success"] = success
	e.Payload["duration_ms"] = durationMs
	return e
}

// AgentFailedEvent creates an agent.failed event with error message.
func AgentFailedEvent(agentID, tenantID, traceID, errMsg string) *Event {
	e := NewEvent(AgentFailed, agentID, tenantID, traceID)
	e.Payload["error"] = errMsg
	return e
}

// PolicyBlockedEvent creates a policy.blocked event.
func PolicyBlockedEvent(agentID, tenantID, traceID, action, reason string) *Event {
	e := NewEvent(PolicyBlocked, agentID, tenantID, traceID)
	e.Payload["action"] = action
	e.Payload["reason"] = reason
	return e
}

// ToolCalledEvent creates an agent.tool_called event.
func ToolCalledEvent(agentID, tenantID, traceID, toolName, inputSummary string) *Event {
	e := NewEvent(ToolCalled, agentID, tenantID, traceID)
	e.Payload["tool_name"] = toolName
	e.Payload["input_summary"] = inputSummary
	return e
}

// GraphWriteEvent creates a sif.graph_write event.
func GraphWriteEvent(agentID, tenantID, traceID, nodeType, nodeID, operation string) *Event {
	e := NewEvent(GraphWrite, agentID, tenantID, traceID)
	e.Payload["node_type"] = nodeType
	e.Payload["node_id"] = nodeID
	e.Payload["operation"] = operation
	return e
}
